graphql_compiler/global_utils.py

```diff
@@ -17,13 +20,27 @@ class PropertyPath(NamedTuple):
     field_name: str
 
 
+QueryStringWithParametersT = TypeVar(
+    "QueryStringWithParametersT", bound="QueryStringWithParameters"
+)
+ASTWithParametersT = TypeVar("ASTWithParametersT", bound="ASTWithParameters")
+
+
 @dataclass
 class QueryStringWithParameters:
     """A query string and parameters that validate against the query."""
 
     query_string: str
     parameters: Dict[str, Any]
 
+    @classmethod
+    def from_ast_with_parameters(
+        cls: Type[QueryStringWithParametersT], ast_with_params: "ASTWithParameters"
+    ) -> QueryStringWithParametersT:
```

Comment:
Can you not use a string here instead, i.e. "QueryStringWithParameters"? The main reason I prefer a string here is for future consistency; I would argue that other developers are likely to use a string here since this what is suggested by the top StackOverlow post when you search for "return type of a method is the same as class python".

https://stackoverflow.com/questions/33533148/how-do-i-specify-that-the-return-type-of-a-method-is-the-same-as-the-class-itsel

Reply:
The two options on the table are the following:
```python
# option 1 -- the current code
QueryStringWithParametersT = TypeVar(
    "QueryStringWithParametersT", bound="QueryStringWithParameters"
)
@classmethod
def from_ast_with_parameters(
    cls: Type[QueryStringWithParametersT], ast_with_params: "ASTWithParameters"
) -> QueryStringWithParametersT:
    pass

# option 2 -- your suggestion
@classmethod
def from_ast_with_parameters(
    cls, ast_with_params: "ASTWithParameters"
) -> "QueryStringWithParameters":
    pass
```

Despite the similarity, they say very different things. The former says:
- "Given a type, which is a subtype of `QueryStringWithParameters`, return an instance of that type".

The latter says:
- "Given a type, return an instance of `QueryStringWithParameters`. The input type is not (and likely, cannot safely be) used anywhere in the type definition or implementation.

I think the former is preferable, because it continues to be safe in the presence of subclasses. Also, see below.